dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java

```diff
@@ -0,0 +1,459 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package io.openliberty.microprofile.openapi20.internal;
+
+import static java.util.stream.Collectors.toList;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.container.service.app.deploy.ModuleInfo;
+import com.ibm.ws.kernel.productinfo.ProductInfo;
+
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIAppConfigProvider;
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIServerXMLConfig;
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIServerXMLConfig.ConfigMode;
+import io.openliberty.microprofile.openapi20.internal.services.ModuleSelectionConfig;
+import io.openliberty.microprofile.openapi20.internal.utils.Constants;
+import io.openliberty.microprofile.openapi20.internal.utils.MessageConstants;
+
+/**
+ * Handles reading the merge include/exclude configuration properties and indicating whether a particular module should be included or excluded.
+ */
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
+public class ModuleSelectionConfigImpl implements ModuleSelectionConfig, OpenAPIAppConfigProvider.OpenAPIAppConfigListener {
+
+    private static final TraceComponent tc = Tr.register(ModuleSelectionConfigImpl.class);
+
+    private volatile ConfigValues configValues = null;
+
+    @Reference
+    private OpenAPIAppConfigProvider configFromServerXMLProvider;
+
+    @Activate
+    public void activate() {
+        configFromServerXMLProvider.registerAppConfigListener(this);
+    }
+
+    @Deactivate
+    public void deactivate() {
+        configFromServerXMLProvider.unregisterAppConfigListener(this);
+    }
+
+    @Override
+    public String toString() {
+        if (configValues == null) {
+            return ("Unconfigured Module Selection Config");
+        }
+
+        ConfigValues config = configValues;
+        StringBuilder sb = new StringBuilder("Module Selection Config[");
```

Comment:
```suggestion
        ConfigValues config = configValues;
        if (config == null) {
            return ("Unconfigured Module Selection Config");
        }

        StringBuilder sb = new StringBuilder("Module Selection Config[");
```

Must only read the `configValues` field once since it could change.